draft-shafranovich-rfc4180-bis.md

```diff
@@ -115,7 +115,7 @@ BE ignored. For example:
    "aaa","bbb", "ccc" CRLF<br/>
    zzz,yyy,xxx
 
-6. Fields containing line breaks (CR or CRLF), double quotes, and commas
+6. Fields containing line breaks (CR or LF), double quotes, or commas
```

Comment:
I think the correct thing would be "CRLF" or "LF", not CR

Reply:
ABNF says `*(TEXTDATA / COMMA / CR / LF / 2DQUOTE)` – not `*(TEXTDATA / COMMA / CRLF / LF / 2DQUOTE)`.

A library that creates a CSV file and has to write a field containing CR (valid MacOS line delimiter) – what should happen then?

Reply:
Ah excellent question. IETF defines line breaks as CRLF only (as per RFC 2046). That is why the original RFC was worded that way. Obviously, we are trying to document the common denominator for the entire Internet, so it would make sense to be more lenient. The relaxation to allow "LF" in addition to "CRLF" came originally the W3C:

https://github.com/w3c/csvw/issues/45

This is based on the approach taken by HTTP standard (RFC 7230, section 3.5). To the best of my knowledge, no IETF specification assumes CR to be a valid line break, since only the legacy MacOS had it and no recent version supports it. I also haven't seen it yet in the wild but would need to do some research.

With that being said, there are several choices here. I am leaning towards allow CRLF or LF as part of the ABNF grammar, and mentioning that CR may also be a valid value. That is kind of what is already implied by the following:

> However, implementors should be aware that some implementations may use other values.

Reply:
I haven't seen CR as a standalone newline character in the wild by myself. It really doesn't seem to play a role. My problem would be the practical side of the story. Of course implementors have the choice to support other values but the RFC should at least allow a clean implementation. Let me try to explain that.

Currently we have:
```
escaped = *(WSP) DQUOTE *(TEXTDATA / COMMA / CR / LF / 2DQUOTE) DQUOTE *(WSP)
non-escaped = *TEXTDATA
TEXTDATA =  %x20-21 / %x23-2B / %x2D-7E
```

By that, a non-escaped field cannot contain `COMMA / CR / LF / 2DQUOTE`. An escaped field can contain these characters. If this is changed to:

```
escaped = *(WSP) DQUOTE *(TEXTDATA / COMMA / [CR]LF / 2DQUOTE) DQUOTE *(WSP)
```

The CR character itself does not become illegal – because it is part of the CRLF sequence. But if the implementor is reading/writing CR and the next character is not LF – what to do? The CR cannot be handled as TEXTDATA and it is also not covered by the new `[CR]LF` sequence rule.

I have the feeling that this could cause some headache for implementors. What is the benefit?

Reply:
I've found something in "Canonicalization and Text Defaults" of HTTP/1.1 [Section 3.7.1 of RFC2616](https://tools.ietf.org/html/rfc2616#section-3.7.1):

>  When in canonical form, media subtypes of the "text" type use CRLF as
>  the text line break. HTTP relaxes this requirement and allows the
>  transport of text media with plain CR or LF alone representing a line
>  break when it is done consistently for an entire entity-body. HTTP
>  applications MUST accept CRLF, bare CR, and bare LF as being
>  representative of a line break in text media received via HTTP.

I think, we should do similar.

Reply:
Looks like this carried over into RFC 7231. I adjusted the draft to reflect all three possible values, and re-did the ABNF to match. Can you take a look?

Reply:
Looks good to me! I'm only wondering if `linebreak = CR / LF / CRLF` the correct/best ABNF syntax for this. But if not, that would be a differed PR.